File: sdk/go/checkpoint.go
package agenttrace

import (
	"crypto/sha256"
	"encoding/hex"
	"io"
	"os"
	"os/exec"
	"strings"
	"time"

	"github.com/google/uuid"
)

// CheckpointType represents the type of checkpoint.
type CheckpointType string

const (
	CheckpointTypeManual   CheckpointType = "manual"
	CheckpointTypeAuto     CheckpointType = "auto"
	CheckpointTypeToolCall CheckpointType = "tool_call"
	CheckpointTypeError    CheckpointType = "error"
	CheckpointTypeMilestone CheckpointType = "milestone"
	CheckpointTypeRestore  CheckpointType = "restore"
)

// CheckpointOptions holds options for creating a checkpoint.
type CheckpointOptions struct {
	Name           string
	Type           CheckpointType
	ObservationID  string
	Description    string
	Files          []string
	IncludeGitInfo bool
}

// CheckpointInfo contains information about a created checkpoint.
type CheckpointInfo struct {
	ID             string
	Name           string
	Type           CheckpointType
	TraceID        string
	ObservationID  string
	GitCommitSha   string
	GitBranch      string
	FilesChanged   []string
	TotalFiles     int
	TotalSizeBytes int64
	CreatedAt      time.Time
}

// GitInfo contains git repository information.
type GitInfo struct {
	CommitSha string
	Branch    string
	RepoURL   string
}

// Checkpoint creates a checkpoint for this trace.
func (t *Trace) Checkpoint(opts CheckpointOptions) *CheckpointInfo {
	checkpointID := uuid.New().String()
	now := time.Now().UTC()

	if opts.Type == "" {
		opts.Type = CheckpointTypeManual
	}

	// Include git info by default
	if opts.IncludeGitInfo || opts.IncludeGitInfo == false && len(opts.Files) == 0 {
		opts.IncludeGitInfo = true
	}

	var gitInfo GitInfo
	if opts.IncludeGitInfo {
		gitInfo = getGitInfo()
	}

	// Calculate file info
	filesChanged := opts.Files
	if filesChanged == nil {
		filesChanged = []string{}
	}
	totalFiles := len(filesChanged)
	var totalSizeBytes int64

	filesSnapshot := make(map[string]map[string]interface{})
	for _, filePath := range filesChanged {
		if info, err := os.Stat(filePath); err == nil {
			totalSizeBytes += info.Size()

			if f, err := os.Open(filePath); err == nil {
				h := sha256.New()
				io.Copy(h, f)
				f.Close()
				hash := hex.EncodeToString(h.Sum(nil))
				filesSnapshot[filePath] = map[string]interface{}{
					"size": info.Size(),
					"hash": hash,
				}
			}
		}
	}

	// Send to API
	if t.client.Enabled() {
		t.client.addEvent(map[string]any{
			"type": "checkpoint-create",
			"body": map[string]any{
				"id":             checkpointID,
				"traceId":        t.id,
				"observationId":  opts.ObservationID,
				"name":           opts.Name,
				"description":    opts.Description,
				"type":           string(opts.Type),
				"gitCommitSha":   gitInfo.CommitSha,
				"gitBranch":      gitInfo.Branch,
				"gitRepoUrl":     gitInfo.RepoURL,
				"filesSnapshot":  filesSnapshot,
				"filesChanged":   filesChanged,
				"totalFiles":     totalFiles,
				"totalSizeBytes": totalSizeBytes,
				"timestamp":      now.Format(time.RFC3339Nano),
			},
		})
	}

	return &CheckpointInfo{
		ID:             checkpointID,
		Name:           opts.Name,
		Type:           opts.Type,
		TraceID:        t.id,
		ObservationID:  opts.ObservationID,
		GitCommitSha:   gitInfo.CommitSha,
		GitBranch:      gitInfo.Branch,
		FilesChanged:   filesChanged,
		TotalFiles:     totalFiles,
		TotalSizeBytes: totalSizeBytes,
		CreatedAt:      now,
	}
}

func getGitInfo() GitInfo {
	var info GitInfo

	if out, err := exec.Command("git", "rev-parse", "HEAD").Output(); err == nil {
		info.CommitSha = strings.TrimSpace(string(out))
	}

	if out, err := exec.Command("git", "rev-parse", "--abbrev-ref", "HEAD").Output(); err == nil {
		info.Branch = strings.TrimSpace(string(out))
	}

	if out, err := exec.Command("git", "config", "--get", "remote.origin.url").Output(); err == nil {
		info.RepoURL = strings.TrimSpace(string(out))
	}

	return info
}

File: sdk/typescript/src/checkpoint.ts
/**
 * Checkpoint functionality for AgentTrace.
 *
 * Checkpoints allow you to create snapshots of code state during agent execution,
 * enabling restoration and debugging of agent sessions.
 */

import { execSync } from "child_process";
import { statSync, readFileSync, existsSync } from "fs";
import { createHash } from "crypto";
import type { AgentTrace } from "./client";

export type CheckpointType = "manual" | "auto" | "tool_call" | "error" | "milestone" | "restore";

export interface CheckpointOptions {
  name: string;
  type?: CheckpointType;
  observationId?: string;
  description?: string;
  files?: string[];
  includeGitInfo?: boolean;
}

export interface CheckpointInfo {
  id: string;
  name: string;
  type: CheckpointType;
  traceId: string;
  observationId?: string;
  gitCommitSha?: string;
  gitBranch?: string;
  filesChanged: string[];
  totalFiles: number;
  totalSizeBytes: number;
  createdAt: Date;
}

export interface GitInfo {
  commitSha?: string;
  branch?: string;
  repoUrl?: string;
}

/**
 * Client for creating and managing checkpoints.
 *
 * @example
 * ```typescript
 * const client = new AgentTrace({ apiKey: "..." });
 * const trace = client.trace({ name: "my-agent" });
 *
 * // Create a checkpoint
 * const cp = trace.checkpoint({
 *   name: "before-edit",
 *   type: "manual",
 *   files: ["src/main.ts", "src/utils.ts"]
 * });
 * ```
 */
export class CheckpointClient {
  private _client: AgentTrace;

  constructor(client: AgentTrace) {
    this._client = client;
  }

  /**
   * Create a new checkpoint.
   */
  create(traceId: string, options: CheckpointOptions): CheckpointInfo {
    const checkpointId = generateId();
    const now = new Date();
    const type = options.type || "manual";

    // Gather git info if requested
    let gitInfo: GitInfo = {};
    if (options.includeGitInfo !== false) {
      gitInfo = this.getGitInfo();
    }

    // Calculate file info
    const filesChanged = options.files || [];
    const totalFiles = filesChanged.length;
    let totalSizeBytes = 0;

    const filesSnapshot: Record<string, { size: number; hash: string }> = {};
    for (const filePath of filesChanged) {
      if (existsSync(filePath)) {
        try {
          const stat = statSync(filePath);
          totalSizeBytes += stat.size;

          const content = readFileSync(filePath);
          const hash = createHash("sha256").update(content).digest("hex");
          filesSnapshot[filePath] = { size: stat.size, hash };
        } catch {
          // Ignore errors
        }
      }
    }

    // Send to API
    if (this._client.enabled) {
      this._client._addEvent({
        type: "checkpoint-create",
        body: {
          id: checkpointId,
          traceId,
          observationId: options.observationId,
          name: options.name,
          description: options.description,
          type,
          gitCommitSha: gitInfo.commitSha,
          gitBranch: gitInfo.branch,
          gitRepoUrl: gitInfo.repoUrl,
          filesSnapshot,
          filesChanged,
          totalFiles,
          totalSizeBytes,
          timestamp: now.toISOString(),
        },
      });
    }

    return {
      id: checkpointId,
      name: options.name,
      type,
      traceId,
      observationId: options.observationId,
      gitCommitSha: gitInfo.commitSha,
      gitBranch: gitInfo.branch,
      filesChanged,
      totalFiles,
      totalSizeBytes,
      createdAt: now,
    };
  }

  /**
   * Get current git repository info.
   */
  getGitInfo(): GitInfo {
    const result: GitInfo = {};

    try {
      result.commitSha = execSync("git rev-parse HEAD", { encoding: "utf-8", timeout: 5000 }).trim();
    } catch {
      // Not in a git repo or git not available
    }

    try {
      result.branch = execSync("git rev-parse --abbrev-ref HEAD", { encoding: "utf-8", timeout: 5000 }).trim();
    } catch {
      // Ignore
    }

    try {
      result.repoUrl = execSync("git config --get remote.origin.url", { encoding: "utf-8", timeout: 5000 }).trim();
    } catch {
      // Ignore
    }

    return result;
  }
}

/**
 * Create a checkpoint scope for automatic tracking.
 */
export async function withCheckpoint<T>(
  client: AgentTrace,
  traceId: string,
  options: CheckpointOptions,
  fn: (checkpoint: CheckpointInfo) => Promise<T>
): Promise<T> {
  const cpClient = new CheckpointClient(client);
  const cp = cpClient.create(traceId, options);
  return fn(cp);
}

function generateId(): string {
  if (typeof crypto !== "undefined" && crypto.randomUUID) {
    return crypto.randomUUID();
  }
  return "xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx".replace(/[xy]/g, (c) => {
    const r = (Math.random() * 16) | 0;
    const v = c === "x" ? r : (r & 0x3) | 0x8;
    return v.toString(16);
  });
}
